File: Leetcode100/leetcode100_python/PathSumIII2.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        def dfs(root, sumList):
            if not root:
                return 0
            sumList = [num + root.val for num in sumList]
            sumList.append(root.val)
            return sumList.count(sum) + dfs(root.left, sumList) + dfs(root.right, sumList)
        return dfs(root, [])

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    sum = int(input())
    root = TreeNode(nums[0])
    queue = [root]
    for i in range(1, len(nums), 2):
        cur = queue.pop(0)
        if nums[i] != -1:
            cur.left = TreeNode(nums[i])
            queue.append(cur.left)
        if nums[i + 1] != -1:
            cur.right = TreeNode(nums[i + 1])
            queue.append(cur.right)
    solution = Solution()
    print(solution.pathSum(root, sum))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/path_sum_III/PathSumIII2.java
package main.java.com.jiang.leetcode100.path_sum_III;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Scanner;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution2 {
    public int pathSum(TreeNode root, int sum) {
        if(root == null) return 0;
        List<Integer> sumList = new ArrayList<>();
        return helper(root, sum, sumList);
    }

    private int helper(TreeNode root, int sum, List<Integer> sumList) {
        if(root == null) return 0;
        List<Integer> temp = new ArrayList<>();
        for(Integer num : sumList) {
            temp.add(num + root.val);
        }
        temp.add(root.val);

        int count = 0;
        for(Integer num : temp) {
            if(num == sum) count++;
        }
        return count + helper(root.left, sum, temp) + helper(root.right, sum, temp);
    }
}

public class PathSumIII2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        int sum = scanner.nextInt();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new  int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        TreeNode root = new TreeNode(nums[0], null, null);
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);

        for(int i = 1; i < nums.length; i += 2) {
            TreeNode cur = queue.poll();
            if(nums[i] != -1) {
                cur.left = new TreeNode(nums[i], null, null);
                queue.add(cur.left);
            }
            if(nums[i + 1] != -1) {
                cur.right = new TreeNode(nums[i + 1], null, null);
                queue.add(cur.right);
            }
        }

        System.out.println(new Solution2().pathSum(root, sum));
    }
}